### ISSUE TITLE
realm: event dependency optimization

### ISSUE THREAD

**Issue by streichler** [StanfordLegion/realm#68]
Right now, all Event preconditions are handled in a generic fashion, waiting for a CPU thread on the dependent operation's node to become aware that the precondition has been satisfied before starting the dependent operation.  In many cases, latency can be reduced by pushing the triggering operation further down the stack.  (For example, a CUDA task dependent on another CUDA task can be started with just an inter-stream fence to ensure ordering.)

The works breaks into four subcategories:
- [ ] Create a way for low-level completion information to be attached to an event and for dependent operations to query that information.
- [ ] Specifically go after the CUDA task -> CUDA task case mentioned above as the first optimized path.
- [ ] Try to think through any potential deadlock scenarios that come from starting the dependent task before its precondition is done executing.  (For example, it seems prudent to wait until the precondition has finished issuing work and is just waiting for deferred stuff to complete.)
- [ ] Figure out how we want to show this in profiles.  Are we ok with dependency edges going backwards in time?
